### PR TITLE
MergeTree: Deprecate the PropertyManager Class and Its Exposure

### PR DESCRIPTION
**Pull request by anthony-murphy**
Deprecate the PropertyManager Class and Its Exposure

The `PropertyManager` class, along with the `propertyManager` properties and `addProperties` functions on segments and intervals, are not intended for external use.
These elements will be removed in a future release for the following reasons:
 * There are no scenarios where their direct use is needed.
 * Using them directly will result in eventual consistency problems.
 * Upcoming features will require modifications to these mechanisms.

I've removed most internal usages where the deprecated things are not necessary, to make later modifications easier. Additionally, we no longer instantiate PropertyManager class unless is it is needed. 

[AB#12580](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/12580)

### CODE REVIEW

File: packages/dds/sequence/src/intervals/intervalUtils.ts [hunk 1]
```diff
@@ -159,11 +160,16 @@ export interface ISerializedInterval {
...
+	/**
+	 * @deprecated - This property should not be used externally and will be removed in a subsequent release.
+	 */
+	// eslint-disable-next-line import/no-deprecated
 	propertyManager: PropertiesManager;
 	/***/
 	serialize(): ISerializedInterval;
-	/***/
+	/**
+	 * @deprecated - This property should not be used externally and will be removed in a subsequent release.
```
Comment (anthony-murphy) [inline packages/dds/sequence/src/intervals/intervalUtils.ts:171]:
```suggestion
	 * @deprecated - This function should not be used externally and will be removed in a subsequent release.
```

File: packages/dds/merge-tree/src/localReference.ts [hunk 2]
```diff
@@ -125,7 +132,7 @@ class LocalReference implements LocalReferencePosition {
 		this.offset = offset;
 	}
 
-	public isLeaf(): boolean {
+	public isLeaf(): this is ISegment {
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/localReference.ts:135]: vs code was going crazy with errors without this, but the cmd line build was fine. the change is safe, just odd it had to be made

File: .changeset/fancy-books-smell.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+---
+"section": "deprecation"
+---
+Deprecate the PropertyManager Class and Its Exposure
+
+The `PropertyManager` class, along with the `propertyManager` properties and `addProperties` functions on segments and intervals, are not intended for external use.
+These elements will be removed in a future release for the following reasons:
+  * There are no scenarios where their direct use is needed.
+	* Using them directly will result in eventual consistency problems.
+	* Upcoming features will require modifications to these mechanisms.
```
Comment (anthony-murphy) [inline .changeset/fancy-books-smell.md:16]:
```suggestion
 * Using them directly will result in eventual consistency problems.
 * Upcoming features will require modifications to these mechanisms.
```

File: .changeset/fancy-books-smell.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+"@fluid-experimental/sequence-deprecated": minor
+---
+---
+"section": "deprecation"
+---
+Deprecate the PropertyManager Class and Its Exposure
+
+The `PropertyManager` class, along with the `propertyManager` properties and `addProperties` functions on segments and intervals, are not intended for external use.
+These elements will be removed in a future release for the following reasons:
+  * There are no scenarios where their direct use is needed.
```
Comment (anthony-murphy) [inline .changeset/fancy-books-smell.md:14]:
```suggestion
 * There are no scenarios where their direct use is needed.
```

File: .changeset/fancy-books-smell.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+---
+"section": "deprecation"
+---
+Deprecate the PropertyManager Class and Its Exposure
+
+The `PropertyManager` class, along with the `propertyManager` properties and `addProperties` functions on segments and intervals, are not intended for external use.
+These elements will be removed in a future release for the following reasons:
+ * There are no scenarios where their direct use is needed.
+ * Using them directly will result in eventual consistency problems.
+ * Upcoming features will require modifications to these mechanisms.
```
Comment (tylerbutler) [inline .changeset/fancy-books-smell.md:16]:
```suggestion
The PropertyManager class and related functions and properties are deprecated

The `PropertyManager` class, along with the `propertyManager` properties and `addProperties` functions on segments and intervals, are not intended for external use.
These elements will be removed in a future release for the following reasons:

 * There are no scenarios where their direct use is needed.
 * Using them directly will cause eventual consistency problems.
 * Upcoming features will require modifications to these mechanisms.
```

**Review by tylerbutler — APPROVED** [review]
nits on the changeset; didn't review the code.

File: packages/dds/sequence/src/intervalCollection.ts [hunk 6]
```diff
@@ -304,7 +304,7 @@ export class LocalIntervalCollection<TInterval extends ISerializableInterval> {
 						"Adding an interval that belongs to another interval collection is not permitted",
 					);
 				}
-				interval.addProperties(props);
+				interval.properties = addProperties(interval.properties, props);
```
Comment (jzaffiro) [inline packages/dds/sequence/src/intervalCollection.ts:307]: Is there a signature of `addProperties` that takes more than one `PropertySet` argument?

**Review by jzaffiro — APPROVED** [review]

File: packages/dds/sequence/src/intervalCollection.ts (hunk 6 above)
Comment (anthony-murphy) [reply packages/dds/sequence/src/intervalCollection.ts:307]: no currently, but the method is internal, so we can modify it if needed

**Review by Josmithr — APPROVED** [review]

File: packages/dds/sequence/src/intervalCollection.ts (hunk 6 above)
Comment (jzaffiro) [reply packages/dds/sequence/src/intervalCollection.ts:307]: Is this code expected to work then? Or do you plan to modify it as a part of this task?

**Pull request merged**